### PR TITLE
Improved nested OP_REQUIRES_OK macros.

### PR DESCRIPTION
**Pull request by MichaelBroughton**
The `OP_REQUIRES_OK` macros have been flaking on ocassion when used in functions that get called in parallel. This update adds a new macro to sychronize the threads' local status with a global status iff there is a local failure on that thread.

### CODE REVIEW

File: tensorflow_quantum/core/ops/parse_context.h [hunk 1]
```diff
@@ -16,6 +16,15 @@ limitations under the License.
 #ifndef TFQ_CORE_OPS_PARSE_CONTEXT
 #define TFQ_CORE_OPS_PARSE_CONTEXT
 
+// Syncs a threads work status with some global status.
+#define NESTED_FN_STATUS_SYNC(global_status, local_status, global_lock) \
+  if (TF_PREDICT_FALSE(!local_status.ok())) {                           \
```
Comment (jaeyoo) [inline tensorflow_quantum/core/ops/parse_context.h:21]: My first question is about `TF_PREDICT_FALSE`. I saw `global_status` was initialized with `ok`, so I believe that this if block allows `global_status` to be `failure`. So, I expect the `TF_PREDICT_FALSE(!local_status.ok())` means `update if local_status.ok() is not true`. That is, `if (TF_REDICT_TRUE(!local_status.ok())`?

Comment (MichaelBroughton) [reply tensorflow_quantum/core/ops/parse_context.h:21]: The `TF_PREDICT_TRUE/FALSE` macro is something that hints to the CPUs branch predictor that you expect this condition to be the case almost all of the time. Whenwe have `TF_PREDICT_FALSE(!status.ok())` what that means is tell the branch predictor that you think `!status.ok()` will most of the time be false => status.ok() is true most of the time. In the case of not erroring this is what we would expect. If that is still a little confusing the macro `TF_PREDICT_***` itself doesn't actually change the logic of the code, so if you like you can pretend it's not there :)

File: tensorflow_quantum/core/ops/parse_context.h [hunk 2]
```diff
@@ -16,6 +16,15 @@ limitations under the License.
...
+    global_lock.lock();                                                 \
+    global_status = local_status;                                       \
+    global_lock.unlock();                                               \
+    return;                                                             \
```
Comment (zaqqwerty) [inline tensorflow_quantum/core/ops/parse_context.h:25]: Since this is a macro, not a function, should this line be removed?

**Review by zaqqwerty — CHANGES_REQUESTED** [review]
One more question on the macro

File: tensorflow_quantum/core/ops/parse_context.h (hunk 2 above)
Comment (zaqqwerty) [reply tensorflow_quantum/core/ops/parse_context.h:25]: Is it intended to break out of the inner loops?

Comment (MichaelBroughton) [reply tensorflow_quantum/core/ops/parse_context.h:25]: Yes, the macros is intended to break out of the nested function if there was a problem. That's why the return is included.

**Review by zaqqwerty — APPROVED** [review]
LGTM

**Review by jaeyoo — APPROVED** [review]
LGTM too

**Pull request merged**